docs/src/content/docs/en/spikes/tech-spikes-github-graphql-api-about-replacing-existing-rest-api.mdx

```diff
@@ -0,0 +1,435 @@
+---
+title: GitHub GraphQL API about replacing existing REST API
+description: GitHub GraphQL API about replacing existing REST API
+---
+
+## Background
+
+Currently, our GitHub API are all REST API, and the REST API has the problem of over-fetching or multiple requests.
+
+## Expect
+
+We want to use the GraphQL API to retrieve the key information we need. With the data obtained this way, we can calculate related metrics.
+
+## Solutions
+
+### 1、Endpoint
+
+The GraphQL API endpoint is `https://api.github.com/graphql`. All requests must be HTTP `POST` requests with `application/json` encoded bodies.
+
+### 2、Authentication
+
+Pass the token in your GraphQL request using the `Authorization` HTTP header with a value `token <token>`.
+
+### 3、Using GitHub GraphQL to replace REST
+
+#### 3.1. Verify the GitHub token
+
+##### 3.1.1 Replaced API and Method local
+
+- Replaced api
+
+```
+GET：https://api.github.com/octocat
+```
+
+- Method local
+
+```
+GitHubFeignClient.verifyToken
+```
+
+##### 3.1.2 API replaced by
+
+```
+query {
+  viewer {
+    login
+  }
+}
+```
+
+- The meaning of response status 
+
+| response status | mean                |
+|-----------------|---------------------|
+| 200             | Verification passed |
+| 401             | Verification failed |
+
+#### 3.2. Verify the GitHub branch
+
+##### 3.1.1 Replaced API and Method local
+
+- Replaced api
+
+```
+GET：https://api.github.com/repos/{repository}/branches/{branchName}
+```
+
+- Method local
+
+```
+GitHubFeignClient.verifyCanReadTargetBranch
+```
+
+##### 3.1.2 API replaced by
+
+```
+query { 
+  repository(owner: string, name: string) { 
+    ref(qualifiedName: "refs/heads/branchName: string") { 
+      name
+    }
+  }
+}
+```
+
+- Required value and demo
+
+| Required value |  explain    | demo         |
+|----------------|-------------|--------------|
+| owner          | owner       | au-heartbeat |
+| name           | repo        | Heartbeat    |
+| branchName     | branch name | main         |
+
+if you want to get `main` branch for `au-heartbeat/Heartbeat`, you can use as following.
+
+```
+query { 
```

Comment:
Follow the `Test GraphQL API curl` in [this page ](https://au-heartbeat.github.io/Heartbeat/en/spikes/tech-spikes-buildkite-graphql-api-about-replacing-existing-rest-api/ )

give a excutable cmd to show your spike result

Reply:
@BoBoDai  and otheres

Reply:
@BoBoDai are there any other RESTful APIs that can be using **one GraphQL API** to request? just a question

Reply:
@guzhongren Only the getPullRequestListInfo and getPullRequestCommitInfo api, I wrote below can be merged, other api are call alone.